src/hotspot/share/prims/jvm.cpp

```diff
@@ -1363,6 +1363,67 @@ JVM_ENTRY(jobject, JVM_GetStackAccessControlContext(JNIEnv *env, jclass cls))
   return JNIHandles::make_local(THREAD, result);
 JVM_END
 
+class ScopedValueBindingsResolver {
+public:
+  InstanceKlass* Carrier_klass;
+  Method *vthread_run_runnable_method;
+  Method *thread_run_method;
+
+  ScopedValueBindingsResolver(JavaThread* THREAD) {
+    Klass *k = SystemDictionary::resolve_or_fail(vmSymbols::jdk_incubator_concurrent_ScopedValue_Carrier(), true, THREAD);
+    Carrier_klass = InstanceKlass::cast(k);
+
+    vthread_run_runnable_method = vmClasses::VirtualThread_klass()->find_instance_method
+      (vmSymbols::run_method_name(), vmSymbols::runnable_void_signature(), Klass::PrivateLookupMode::find);
+    guarantee(vthread_run_runnable_method != NULL, "must be");
+
+    thread_run_method = vmClasses::Thread_klass()->find_instance_method
+      (vmSymbols::run_method_name(), vmSymbols::void_method_signature(), Klass::PrivateLookupMode::find);
+    guarantee(thread_run_method != NULL, "must be");
+  }
+};
+
+JVM_ENTRY(jobject, JVM_FindScopedValueBindings(JNIEnv *env, jclass cls))
+  ResourceMark rm(THREAD);
+  GrowableArray<Handle>* local_array = new GrowableArray<Handle>(12);
+  JvmtiVMObjectAllocEventCollector oam;
+
+  bool found = false;
+
+  static ScopedValueBindingsResolver resolver(THREAD);
+
+  // Iterate through Java frames
+  vframeStream vfst(thread);
+  for(; !vfst.at_end(); vfst.next()) {
+    int loc = 0;
+    // get method of frame
+    Method* method = vfst.method();
+
+    Symbol *name = method->name();
+
+    if (method->method_holder() == resolver.Carrier_klass &&
+        (name == vmSymbols::run_method_name() || name == vmSymbols::call_method_name())) {
+      loc = 2;
+    } else if (method == resolver.vthread_run_runnable_method) {
+      loc = 3;
+    } else if (method == resolver.thread_run_method) {
+      loc = 2;
```

Comment:
This depends on how javac numbers locals, right?  It seems a bit fragile.  This is one of the reasons why doPrivileged uses the helper method executePrivileged, so the locals are arguments, giving them predictable offsets.

Reply:
Ah, good point. I'll have a look at doing that.

Reply:
Done.